dynawo/sources/Models/CPP/Controls/Shunts/ModelCentralizedShuntsSectionControl/DYNModelCentralizedShuntsSectionControl.cpp

```diff
@@ -0,0 +1,346 @@
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::evalF(double /*t*/, propertyF_t /*type*/) {
+    /* not need */
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::evalJt(const double /*t*/, const double /*cj*/, SparseMatrix& /*jt*/, const int /*rowOffset*/) {
+    /* not need */
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::evalG(const double t) {
+    if (!isSelf_) {
+      for (int s = 0; s < nbShunts_; ++s) {
+        if (doubleNotEquals(yLocal_[0], (zLocal_[s] - DeadbandsUPu_[s])) &&
+          yLocal_[0] < (zLocal_[s] - DeadbandsUPu_[s]) &&
+          sections0_[s] < SectionsMax_[s]) {
+          gLocal_[0] = ROOT_UP;
+          gLocal_[1] = ROOT_DOWN;
+          gLocal_[2] = ((t - whenUp_) > tNext_ || doubleEquals((t - whenUp_), tNext_)) ? ROOT_UP : ROOT_DOWN;
+          gLocal_[3] = ROOT_DOWN;
+          changingShunt = s;
+          break;
+        } else if (doubleNotEquals(yLocal_[0], (zLocal_[s] + DeadbandsUPu_[s])) &&
+          yLocal_[0] > (zLocal_[s] + DeadbandsUPu_[s]) &&
+          sections0_[s] > SectionsMin_[s]) {
+          gLocal_[0] = ROOT_DOWN;
+          gLocal_[1] = ROOT_UP;
+          gLocal_[2] = ROOT_DOWN;
+          gLocal_[3] = ((t - whenDown_) > tNext_ || doubleEquals((t - whenDown_), tNext_)) ? ROOT_UP : ROOT_DOWN;
+          changingShunt = s;
+          break;
+        } else {
+          gLocal_[0] = ROOT_DOWN;
+          gLocal_[1] = ROOT_DOWN;
+          gLocal_[2] = ROOT_DOWN;
+          gLocal_[3] = ROOT_DOWN;
+        }
+      }
+    } else {
+      for (int s = 0; s < nbShunts_; ++s) {
+        if (doubleNotEquals(yLocal_[0], (zLocal_[s] - DeadbandsUPu_[s])) &&
+          yLocal_[0] < (zLocal_[s] - DeadbandsUPu_[s]) &&
+          sections0_[s] > SectionsMin_[s]) {
+          gLocal_[0] = ROOT_UP;
+          gLocal_[1] = ROOT_DOWN;
+          changingShunt = s;
+          gLocal_[2] = ((t - whenDown_) > tNext_ || doubleEquals((t - whenDown_), tNext_)) ? ROOT_UP : ROOT_DOWN;
+          gLocal_[3] = ROOT_DOWN;
+        } else if (doubleNotEquals(yLocal_[0], (zLocal_[s] + DeadbandsUPu_[s])) &&
+          yLocal_[0] > (zLocal_[s] + DeadbandsUPu_[s]) &&
+          sections0_[s] < SectionsMax_[s]) {
+          gLocal_[0] = ROOT_DOWN;
+          gLocal_[1] = ROOT_UP;
+          gLocal_[2] = ROOT_DOWN;
+          gLocal_[3] = ((t - whenUp_) > tNext_ || doubleEquals((t - whenUp_), tNext_)) ? ROOT_UP : ROOT_DOWN;
+          changingShunt = s;
+        } else {
+          gLocal_[0] = ROOT_DOWN;
+          gLocal_[1] = ROOT_DOWN;
+          gLocal_[2] = ROOT_DOWN;
+          gLocal_[3] = ROOT_DOWN;
+        }
+      }
+    }
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::evalZ(const double t) {
+    if (!isSelf_) {
+      if (gLocal_[0] == ROOT_UP) {
+        whenUp_ = t;
+        whenDown_ = VALDEF;
+      }
+      if (gLocal_[1] == ROOT_UP) {
+        whenDown_ = t;
+        whenUp_ = VALDEF;
+      }
+      if (gLocal_[2] == ROOT_UP) {
+        if (lastTime_ != t) {
```

Comment:
doubleNotEquals

Reply:
another way of doing this is keep the previous section of the shunt (similarly to a pre) and set the value to pre(section)+1. I don't know which one is better though.

Reply:
doubleNotEquals fixed, thanks!